tests/Spec.hs

```diff
@@ -1948,6 +1960,265 @@ main = hspec $ do
+            , BuildStatusChanged (Sha "1b2") "mandatory" (Project.BuildStarted "example.com/mandatory/1b2")
+            , CommentAdded (PullRequestId 1) "bot" "[CI job :yellow_circle:](example.com/required/1b2) started."
+            , BuildStatusChanged (Sha "1b2") "required" Project.BuildSucceeded
+            , BuildStatusChanged (Sha "1b2") "mandatory" Project.BuildSucceeded
+            ]
+          (finalState, actions) = runActionCustom results $ handleEventsTest events state
+        actions `shouldBe`
+          [ AIsReviewer (Username "deckard")
+          , ALeaveComment (PullRequestId 12) "Pull request approved for merge by @deckard, rebasing now."
+          , ATryIntegrate "Merge #12: Twelfth PR\n\n\
+                          \Approved-by: deckard\n\
+                          \Auto-deploy: false\n"
+                          (PullRequestId 12,Branch "refs/pull/12/head",Sha "12a")
+                          []
+                          False
+          , ALeaveComment (PullRequestId 12) "Rebased as 1b2, waiting for CI …"
+          , ALeaveComment (PullRequestId 12) "[CI job :yellow_circle:](example.com/required/1b2) started."
+          , ATryPromote (Branch "tth") (Sha "1b2")
+          , ACleanupTestBranch (PullRequestId 12)
+          ]
+        -- test caveat, in reality, when Promote works,
+        -- the PR is removed from the building list.
+        classifiedPullRequestIds finalState `shouldBe` ClassifiedPullRequestIds
+          { building = [PullRequestId 12] -- not here in a real scenario
+          , failed   = []
+          , approved = []
+          , awaiting = []
+          }
+
+      it "does not merge when one mandatory check out of many fail" $ do
+        let
+          projectState = Project.emptyProjectState
+            { Project.mandatoryChecks = Project.MandatoryChecks (Set.fromList ["required", "mandatory"]) }
+          state
+            = Project.insertPullRequest (PullRequestId 12)
+                (Branch "tth") masterBranch (Sha "12a") "Twelfth PR"  (Username "person")
+            $ projectState
+          results = defaultResults {resultIntegrate = [Right (Sha "1b2")]}
+          events =
+            [ CommentAdded (PullRequestId 12) "deckard" "@bot merge"
+            , CommentAdded (PullRequestId 12) "bot" "Pull request approved for merge, rebasing now."
+            , CommentAdded (PullRequestId 12) "bot" "Rebased as 1b2, waiting for CI …"
+            , BuildStatusChanged (Sha "1b2") "required" (Project.BuildStarted "example.com/required/1b2")
+            , BuildStatusChanged (Sha "1b2") "mandatory" (Project.BuildStarted "example.com/mandatory/1b2")
+            , CommentAdded (PullRequestId 1) "bot" "[CI job :yellow_circle:](example.com/required/1b2) started."
+            , BuildStatusChanged (Sha "1b2") "required" (Project.BuildFailed Nothing)
+            , BuildStatusChanged (Sha "1b2") "mandatory" Project.BuildSucceeded
```

Comment:
Based on the above, my guess would be that this test fails if the success was to come first.